src/passes/Asyncify.cpp

```diff
@@ -1214,6 +1244,222 @@ struct AsyncifyFlow : public Pass {
+    return std::make_unique<AsyncifyAddCatchCounters>();
+  }
+
+  void runOnFunction(Module* module_, Function* func) override {
+    class CountersBuilder : public Builder {
+    public:
+      CountersBuilder(Module& wasm) : Builder(wasm) {}
+      Expression* makeInc(int amount = 1) {
+        return makeGlobalSet(
+          ASYNCIFY_CATCH_COUNTER,
+          makeBinary(AddInt32,
+                     makeGlobalGet(ASYNCIFY_CATCH_COUNTER, Type::i32),
+                     makeConst(int32_t(amount))));
+      }
+      Expression* makeDec(int amount = 1) {
+        return makeGlobalSet(
+          ASYNCIFY_CATCH_COUNTER,
+          makeBinary(SubInt32,
+                     makeGlobalGet(ASYNCIFY_CATCH_COUNTER, Type::i32),
+                     makeConst(int32_t(amount))));
+      };
+    };
+    CountersBuilder builder(*module_);
+    BranchUtils::BranchTargets branchTargets(func->body);
+
+    // with this walker we will find level of "nesting" for each expression
+    // ... - +0
+    // catch
+    //   ... - +1
+    //   catch
+    //     ... - +2
+    std::unordered_map<Expression*, int> expressionNestedLevel;
+    struct NestedLevelWalker
+      : public PostWalker<NestedLevelWalker,
+                          UnifiedExpressionVisitor<NestedLevelWalker>> {
+      std::unordered_map<Expression*, int>* expressionNestedLevel;
+      int nestedLevel = 0;
+
+      static void doStartCatch(NestedLevelWalker* self, Expression** currp) {
+        self->nestedLevel++;
+      }
+
+      static void doEndCatch(NestedLevelWalker* self, Expression** currp) {
+        self->nestedLevel--;
+      }
+
+      static void scan(NestedLevelWalker* self, Expression** currp) {
+        auto curr = *currp;
+        if (curr->_id == Expression::Id::TryId) {
+          auto& catchBodies = curr->cast<Try>()->catchBodies;
+          for (Index i = 0; i < catchBodies.size(); i++) {
+            self->pushTask(doEndCatch, currp);
+            self->pushTask(NestedLevelWalker::scan, &catchBodies[i]);
+            self->pushTask(doStartCatch, currp);
+          }
+          self->pushTask(NestedLevelWalker::scan, &curr->cast<Try>()->body);
+          return;
+        }
+
+        PostWalker<NestedLevelWalker,
+                   UnifiedExpressionVisitor<NestedLevelWalker>>::scan(self,
+                                                                      currp);
+      }
+
+      void visitExpression(Expression* curr) {
+        expressionNestedLevel->insert(std::make_pair<>(curr, nestedLevel));
+      }
+    };
+    NestedLevelWalker nestedLevelWalker;
+    nestedLevelWalker.expressionNestedLevel = &expressionNestedLevel;
+    nestedLevelWalker.walk(func->body);
+
+    // with this walker we will handle those counters:
+    // - entering into catch (= pop)  +1
+    // - return                       -N (nested catch count up to root)
+    // - break                        -N (nested catch count up to label)
+    // - exiting from catch           -1
```

Comment:
I think a simpler approach might be to use a `try-finally` pattern? That would also handle nested try-catch, for example. Wasm doesn't have finally, but you can do `catch_all + rethrow`:
```wat
(try
  (do
    COUNTER++;
  )
  (catch_all
    COUNTER--;
    (rethrow 0)
  )
)
```

Reply:
Hi, I alreay use try/catch_all pattern, for all catch blocks, but return/break can't be handled by try/catch_all, so they are also needed. But, I realized that I need modify only root catch block, because we not interested to count nested blocks. I will try to add lit tests, and refactor this...

Reply:
Oh, right, you also need return/break code too... I forgot `catch_all` is not really `finally`...